### PR TITLE
[Docs] Revamp docs for secrets and distributed jobs

### PR DESCRIPTION
**Pull request by Michaelvll**
This PR includes the following:
1. Make `Environment Variables` page level-1 as the secret management and environments are important to our users.
2. Make distributed jobs level-1 for easier reference.
3. Revamp docs to make it easier to follow.

Find the updated docs of this PR: https://skypilot.readthedocs.io/en/secrets-env-var/running-jobs/environment-variables.html

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: docs/source/running-jobs/environment-variables.rst [hunk 1]
```diff
@@ -1,33 +1,67 @@
...
+  
+- ``--env`` flag in ``sky launch/exec`` :ref:`CLI <cli>` (takes precedence over the above)
 
 .. tip::
 
   If an environment variable is required to be specified with `--env` during
-  ``sky launch/exec``, you can set it to ``null`` in task YAML to raise an
+  ``sky launch/exec``, you can set it to empty or ``null`` in task YAML to raise an
   error when it is forgotten to be specified. For example, the ``WANDB_API_KEY``
   and ``HF_TOKEN`` in the following task YAML:
```
Comment (concretevitamin) [inline docs/source/running-jobs/environment-variables.rst:35]: Can we rephrase like this?

To mark an environment variable as **required** and make SkyPilot forcefully check its existence (errors out if unspecified), set it to an empty string or ``null`` in the task YAML. For example, ``WANDB_API_KEY``
  and ``HF_TOKEN`` in the following task YAML are marked as required:

File: docs/source/running-jobs/environment-variables.rst [hunk 2]
```diff
@@ -1,33 +1,67 @@
...
 
 The ``file_mounts``, ``setup``, and ``run`` sections of a task YAML can access the variables via the ``${MYVAR}`` syntax.
 
+.. _passing-secrets:
+
+Passing secrets
+~~~~~~~~~~~~~~~
+
+We recommend passing secrets to any node(s) executing your task by first making
+it available in your current shell, then using ``--env`` to pass it to SkyPilot:
```
Comment (concretevitamin) [inline docs/source/running-jobs/environment-variables.rst:52]:
```suggestion
We recommend passing secrets to any node(s) executing your task by first making
it available in your current shell, then using ``--env SECRET`` to pass it to SkyPilot:
```

File: docs/source/running-jobs/environment-variables.rst [hunk 3]
```diff
@@ -1,33 +1,67 @@
...
+
+.. code-block:: console
+
+  $ sky launch -c mycluster --env HF_TOKEN --env WANDB_API_KEY task.yaml
+  $ sky exec mycluster --env WANDB_API_KEY task.yaml
+
+.. tip::
+
+  When value is not specified for ``--env HF_TOKEN``, SkyPilot reads it from
+  local environment variables
```
Comment (concretevitamin) [inline docs/source/running-jobs/environment-variables.rst:62]:
```suggestion
  When the value is not specified (e.g., ``--env HF_TOKEN``), SkyPilot reads it from
  local environment variables.
```

Comment (concretevitamin) [reply docs/source/running-jobs/environment-variables.rst:62]: How about keeping this sentence & append the above ?
```
   In other words, you do not need to pass the value directly such as ``--env
   WANDB_API_KEY=1234``.
```

File: docs/source/running-jobs/environment-variables.rst [hunk 4]
```diff
@@ -77,40 +111,24 @@ For example, this is useful for passing secrets (see below) or passing configura
...
 SkyPilot environment variables
 ------------------------------------------------------------------
 
-SkyPilot exports these environment variables for a task's execution. ``setup``
-and ``run`` stages have different environment variables available.
+SkyPilot exports several predefined environment variables for a task's execution, which
+are useful for frameworks that require information about the cluster or task, such as
+torch.distributed, OpenMPI, etc. See examples in :ref:`dist-jobs` and :ref:`managed-jobs`.
+
+``setup`` and ``run`` stages have different environment variables available:
```
Comment (concretevitamin) [inline docs/source/running-jobs/environment-variables.rst:124]: Should L234-240 be moved to this overview section?

File: docs/source/running-jobs/environment-variables.rst [hunk 5]
```diff
@@ -77,40 +111,24 @@ For example, this is useful for passing secrets (see below) or passing configura
...
+.. _sky-env-vars:
 
 SkyPilot environment variables
 ------------------------------------------------------------------
 
-SkyPilot exports these environment variables for a task's execution. ``setup``
-and ``run`` stages have different environment variables available.
+SkyPilot exports several predefined environment variables for a task's execution, which
+are useful for frameworks that require information about the cluster or task, such as
+torch.distributed, OpenMPI, etc. See examples in :ref:`dist-jobs` and :ref:`managed-jobs`.
```
Comment (concretevitamin) [inline docs/source/running-jobs/environment-variables.rst:122]:
```suggestion
SkyPilot exports several predefined environment variables made available during a task's execution. These variables contain information about the current cluster or task, which can be useful for distributed frameworks such as
torch.distributed, OpenMPI, etc. See examples in :ref:`dist-jobs` and :ref:`managed-jobs`.
```

File: docs/source/running-jobs/environment-variables.rst (hunk 4 above)
Comment (concretevitamin) [inline docs/source/running-jobs/environment-variables.rst:124]:
```suggestion
The ``setup`` and ``run`` stages can access different sets of SkyPilot environment variables:
```

File: docs/source/running-jobs/environment-variables.rst [hunk 6]
```diff
@@ -120,9 +138,15 @@ Environment variables for ``setup``
      - Rank (an integer ID from 0 to :code:`num_nodes-1`) of the node being set up.
      - 0
    * - ``SKYPILOT_SETUP_NODE_IPS``
-     - A string of IP addresses of the nodes in the cluster with the same order as the node ranks, where each line contains one IP address. Note that this is not necessarily the same as the nodes in ``run`` stage, as the ``setup`` stage runs on all nodes of the cluster, while the ``run`` stage can run on a subset of nodes.
-     - 1.2.3.4
-       3.4.5.6
+     - A string of IP addresses of the nodes in the cluster with the same order as the node ranks, where each line contains one IP address.
+     
+       Note that this is not necessarily the same as the nodes in ``run`` stage, as the ``setup`` stage runs on all nodes of the cluster, while the ``run`` stage can run on a subset of nodes.
```
Comment (concretevitamin) [inline docs/source/running-jobs/environment-variables.rst:143]:
```suggestion
     
       Note that this is not necessarily the same as the nodes in ``run`` stage: the ``setup`` stage runs on all nodes of the cluster, while the ``run`` stage can run on a subset of nodes.
```

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll. Some comments.

File: docs/source/running-jobs/distributed-jobs.rst [hunk 7]
```diff
@@ -66,47 +65,32 @@ SkyPilot exposes these environment variables that can be accessed in a task's ``
...
-When using ``sky launch`` to launch a multi-node task on **a new cluster**, the following happens in sequence:
+When using ``sky launch`` to launch a multi-node task, the following happens in sequence:
 
-1. Nodes are provisioned. (barrier)
+1. Provision (if not exists) or verify (if exists) a multi-node cluster. (gang-scheduled)
 2. Workdir/file_mounts are synced to all nodes. (barrier)
 3. ``setup`` commands are executed on all nodes. (barrier)
 4. ``run`` commands are executed on all nodes.
 
-Launching a multi-node task (existing cluster)
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply docs/source/running-jobs/distributed-jobs.rst]: Ahh, just realized that there is another purpose for the two sections, which is to show that it is able to launch on a subset of nodes.

File: docs/source/running-jobs/distributed-jobs.rst [hunk 8]
```diff
@@ -66,47 +65,32 @@ SkyPilot exposes these environment variables that can be accessed in a task's ``
...
-    3p`.
-
-  - To manipulate these IP addresses, you can also store them to a file in the
-    :code:`run` command with :code:`echo $SKYPILOT_NODE_IPS >> ~/sky_node_ips`.
+- :code:`SKYPILOT_NUM_NODES`: number of nodes reserved for the task, specified by ``num_nodes: <n>``, 
+  which is equivalent to :code:`echo "$SKYPILOT_NODE_IPS" | wc -l`
 - :code:`SKYPILOT_NUM_GPUS_PER_NODE`: number of GPUs reserved on each node to execute the
   task; the same as the count in ``accelerators: <name>:<count>`` (rounded up if a fraction).
 
+See :ref:`env-vars` for more details.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply docs/source/running-jobs/distributed-jobs.rst]: Good point! Changed. Thanks!

File: docs/source/running-jobs/environment-variables.rst (hunk 1 above)
Comment (Michaelvll) [reply docs/source/running-jobs/environment-variables.rst]: Good point! Updated.

File: docs/source/running-jobs/environment-variables.rst (hunk 4 above)
Comment (Michaelvll) [reply docs/source/running-jobs/environment-variables.rst]: Done. Moved the lines.

File: docs/source/running-jobs/environment-variables.rst [hunk 9]
```diff
@@ -58,8 +58,9 @@ it available in your current shell, then using ``--env`` to pass it to SkyPilot:
 
 .. tip::
 
-  When value is not specified for ``--env HF_TOKEN``, SkyPilot reads it from
-  local environment variables
+  You do not need to pass the value directly such as ``--env
+  WANDB_API_KEY=1234``. When the value is not specified (e.g., ``--env HF_TOKEN``),
```
Comment (concretevitamin) [inline docs/source/running-jobs/environment-variables.rst:62]:
```suggestion
  WANDB_API_KEY=1234``. When the value is not specified (e.g., ``--env WANDB_API_KEY``),
```

**Review by concretevitamin — APPROVED** [review]
LGTM

**Pull request merged**